File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    private boolean isValid(long t, int n, int[] times){
        long c = 0;
        for(int time : times){
            c += t / time;
        }
        return c >= n;
    }
    
    public long solution(int n, int[] times) {
        long start = 1;
        long end = 100000000000000L;
        while(end > start){
            long t = (start + end) / 2;
            
            if(isValid(t, n, times)){
                end = t;
            } else {
                start = t + 1;
            }
        }
        
        return start;
    }
}
/*
가능한 최대의 시간에서 최소의 시간을 찾는다
최소의 조건은 뭐지?

*/

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    left, right = 1, max(times) * n
    
    while left <= right:
        mid = (left + right) // 2
        people = 0        
        # 이게 왜 가능한 인워수지?
        # 해당 시간에 받을 수 있는 손님의 수
        for time in times:
            people += mid // time
            if people >= n: break
        
        if people >= n:
            answer = mid
            right = mid - 1
            
        elif people < n:
            left = mid + 1
            
    return answer

